File: nvpl_blas/c/strmm.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_strmm
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 2;
    nvpl_int_t N = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    float alpha = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transa = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_SIDE side = CblasLeft;
    enum CBLAS_DIAG diag = CblasNonUnit;
    float * A = NULL;
    float * B = NULL;

    printf("\nExample: cblas_strmm for the triangular matrix-matrix multiplication\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", order=%c"
            ", side=%c, uplo=%c, transa=%c, diag=%c\n", (int64_t)M, (int64_t)N, (int64_t)lda, (int64_t)ldb,
            order_to_char(order), side_to_char(side), uplo_to_char(uplo), transpose_to_char(transa), diag_to_char(diag));
    printf("           alpha=%g\n", alpha);

    nvpl_int_t rowsA = (side == CblasLeft) ? M : N;
    nvpl_int_t colsA = rowsA;
    nvpl_int_t rowsB = M;
    nvpl_int_t colsB = N;

    // allocate memory
    A = (float *)malloc(lda * colsA  * sizeof(float));
    B = (float *)malloc(ldb * colsB  * sizeof(float));

    // fill data
    fill_smatrix(A, rowsA, colsA, lda, order, UPLO_TO_FILL_MODE(uplo), diag);
    fill_smatrix(B, rowsB, colsB, ldb, order, Full, CblasNonUnit);

    // print input data
    print_smatrix(A, rowsA, colsA, lda, "A", order);
    print_smatrix(B, rowsB, colsB, ldb, "B", order);

    // call cblas_strmm
    cblas_strmm(order, side, uplo, transa, diag, M, N, alpha, A, lda, B, ldb);

    // print result
    printf("\nAfter strmm operation:\n");
    print_smatrix(B, rowsB, colsB, ldb, "B", order);

    // release memory
    free(A);
    free(B);
    return 0;
}

File: nvpl_blas/f77/strmm.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     STRMM 
!
!******************************************************************************
      program   STRMM_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1     side, uplo, transa, diag
      parameter       (side='L', uplo='U', transa='N', diag='N')
      integer         m, n, i, j
      integer         lda, ldb
      parameter       (m=2, n=2,
     &                 lda=2, ldb=2)
      real            alpha
      parameter       (alpha=1.0)
      real            a(lda,m), b(ldb,n)

*       External Subroutines
      external        print_smatrix, fill_smatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, m, n, lda, ldb, side, uplo, transa, diag
      print 101, alpha
      call fill_smatrix(a,m,m,lda,uplo,diag)
      call fill_smatrix(b,m,n,ldb,'F','N')
      call print_smatrix(a,m,m,lda,'A')
      call print_smatrix(b,m,n,ldb,'B')
*
*      Call STRMM subroutine
      call STRMM(side,uplo,transa,diag,m,n,alpha,a,lda,b,ldb)
*
      print*
      print 102, 'After STRMM operation:'
      call print_smatrix(b,m,n,ldb,'B')

      stop
 99   format('Example: STRMM for the triangular matrix-matrix'
     &       ' multiplication')
 100  format('#### args: m=',i1,', n=',i1,', lda=',i1,', ldb=',i1,
     &       ', side=',a1,', uplo=',a1,', transa=',a1,', diag=',a1)
 101  format(11x,'alpha=',f3.1)
 102  format(A)
      end
